### DISCUSSION TITLE
Welcome to c-code-style Discussions!

### DISCUSSION THREAD

**Discussion by MaJerle** [MaJerle/c-code-style#7] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Mango0x45** [comment]
What is the rationale behind not using stdbool.h?

**Reply by MaJerle** [reply]
Where did you see this?

**Reply by Mango0x45** [reply]
Sorry, that was a typo haha. I meant `stdbool.h`

**Reply by MaJerle** [reply]
There is no such thing as true and false in C language.

**Reply by Mango0x45** [reply]
Why does that matter?

**Reply by MaJerle** [reply]
You have asked for a reason, I gave the reason. Case closed :)

**Reply by Mango0x45** [reply]
Yes, but I would like to understand the rationale and benefits. Can I not make the same argument about strings? Strings don’t _actually_ exist in C, they’re just arrays of characters, yet we call them strings, use string functions in `string.h` and `strings.h`, and kind of just pretend like they’re a language feature

**Reply by MaJerle** [reply]
No, not the same. You have functions returning `true` and `false`, but you have functions returning `-1` too, which would be true. So preference is to return `1` or `0` and compare them against NOT if you wish.

I'm not going to argue about strings with you, since it is not the same.

**Reply by filker0** [reply]
Late to the game, but technically, there is both `true` and `false`, but it cannot be stored or passed as a value.
`true` can be either `(N == N)` or `!false`, `false` can be either `(N == N + M)` or `!true`, assuming that `N+M != N`. C treats `false` and `0` as equivalent, and `true` as `not false`, representing it as 1 when stored as the value of a conditional expression.

Compilers know this (or at least the one's I've helped write), and can evaluate this at compile time to produce the right control flow, and to short-cut some logical expressions.

**Comment by EmbeddedLife615** [comment]
邮件已收到，我会尽快回复的！

**Comment by EmbeddedLife615** [comment]
邮件已收到，我会尽快回复的！

**Comment by filker0** [comment]
Honoring the request at the start of this thread:

Greetings. I'm a long time Software Engineer (CS degree in 1981, been a C programmer since around 1982), former compiler developer, and a bit of a computer language nerd.

I was looking for existing C coding standards that I can point someone at without them having to go through the opsec for the program I work on to access our repositories, and I came across this one (amongst others). Though there are several things I disagree with in the rules put forward in @MaJerle's text, overall it's reasonable.

When I'm writing (or part of the team writing) a coding standard, I break it into 2 parts:
* Coding Rules - Practices that are critical to quality software, where violation of these rules must be justified on a case-by-case basis and defended against alternative but functionally equivalent conforming constructs. These rules are backed by examples of how they are essential, and how not following them is non-portable, unsafe or unmaintainable. The standards are written with the knowledge that certain things are sometimes unavoidable, for example using `goto` to exit a nested scope such as several layers into nested loops, in contexts where `break` and some extra code in the enclosing scope blocks is not a practical way to achieve the required results. (You find this a lot in low-level drivers.)
* Style Guide - Practices that promote consistent, easy to read, and maintainable code. This usually has some provision for personal taste, but only within some limits. In recent decades, projects I've been on use tools to detect violations of these guidelines; some reject the code, others warn about the violations, and at least one used a code reformatter to bring the code into conformance each time it is checked in. Violations of these rules are not dangerous, but do cause maintenance headaches because the code is not consistent from module to module.

So there are some style things I do differently (where the open brace of a block is placed; I prefer a more "BSD" style), and none of the standards I work with (including the ones I didn't have a hand in) allow multiple variables to be defined in the same declaration (with some exceptions). Nothing that has a static initializer may share a definition with any other variable.

I've not read the entire set of rules, but these are the two that stood out to me at first.

This is not a criticism, it's just joining the conversation as a professional software engineer.
